Implement merge_gds and check_merge_gds

aofarrel opened the pull request:
Oops! I got a little overeager and accidentally finished it early.

Checker workflow is included, tested, and passes, in spite of #14. ~~Maybe exclude PCA doesn't matter on the sample data?~~

juklucas commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,440 @@
+	Int finalDiskSize = gds_size + sample_size + varInclude_size + addldisk
+
+	# Workaround for optional files
+	Boolean def_sampleInc = defined(sample_include_file)
+	Boolean def_variantInc = defined(variant_include_file)
```
Can you put this above the command section?

Also, probably good to stick to underscore instead of a combination of underscore and camelCase.

aofarrel replied:
`defined()` is a WDL command that doesn't work properly in the command section. Where it's currently located makes it basically a private input variable.

I will fix the casing though.

juklucas commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,440 @@
+	Int gds_size = ceil(size(gds, "GB"))
+	Int sample_size = if defined(sample_include_file) then ceil(size(sample_include_file, "GB")) else 0
+	Int varInclude_size = if defined(variant_include_file) then ceil(size(variant_include_file, "GB")) else 0
+	Int finalDiskSize = gds_size + sample_size + varInclude_size + addldisk
```
varInclude_size should be var_include_size or something with an underscore

aofarrel replied:
Original plan was to stick to _ for stuff in the CWL and camelCase for my own variables but that didn't work.

juklucas commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,440 @@
+	
+	# Estimate disk size required
+	Int gds_size = ceil(size(gds, "GB"))
+	Int sample_size = if defined(sample_include_file) then ceil(size(sample_include_file, "GB")) else 0
```
This will always be really small compared to the GDS, I think you can drop it.

juklucas commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,440 @@
+	# Estimate disk size required
+	Int gds_size = ceil(size(gds, "GB"))
+	Int sample_size = if defined(sample_include_file) then ceil(size(sample_include_file, "GB")) else 0
+	Int varInclude_size = if defined(variant_include_file) then ceil(size(variant_include_file, "GB")) else 0
```
This too can be dropped, I think.

juklucas commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,440 @@
+	}
+
+	# CWL uses a dotproduct scatter; this is the closest WDL equivalent
+	scatter(gds_n_varinc in zip(gds_files, ld_pruning.ld_pruning_output)) {
```
Shouldn't this call be to
```
scatter(gds_n_varinc in zip(ld_pruning.ld_pruning_output, variant_include_file)) {
```
It seems that you are including the original GDS and the LD Pruned GDS not the LD pruned GDS and the variants to include.

juklucas replied:
And if that is the case, then you don't actually need to zip() this.

aofarrel replied:
LD Pruning does not actually output GDS files. LD Pruning outputs RData files saying "here are the places to cut out." It is the subset GDS task who then takes in the GDS files and the LD Pruning RData files and does the actual cutting.

juklucas commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,440 @@
+		File? sample_include_file
+
+		# runtime attributes
+		Int addldisk = 1
```
In general, I'd set this to 20, or something reasonable.

aofarrel replied:
I'm confused. We shouldn't include sample and variant files when calculating disk size, however, we should throw 20 extra gigabytes on top?

aofarrel replied:
We'll go with 5!

juklucas commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,440 @@
+		preemptibles: "${preempt}"
+	}
+	output {
+		File ld_pruning_output = glob("*.RData")[0]
```
We might want to find a better name than ld_pruning_output.

aofarrel replied:
Any suggestions? You're right, it is redundant especially as it is usually called as ld_pruning.ld_pruning_output. I can't just name it output though.

aofarrel replied:
Julian says he was tired and takes this back.

juklucas commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,440 @@
+
+		def py_rootPlusChr(py_filename):
+			'''
+			Similar to split_n_space in vcf-to-gds but not equivalent
+			The CWL uses nameroot for this, but I think the WDL needs the full path
```
I think we can get rid of this comment.

aofarrel replied:
no.

juklucas commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,440 @@
+	input {
+		Pair[File, File] gds_n_varinc  # [gds, variant_include_file]
+		String? out_prefix
+		File? sample_include_file
```
variant_include_file should be an optional variable here (not part of a pair)

aofarrel replied:
https://github.com/DataBiosphere/analysis_pipeline_WDL/blob/implement-merge-gds/documentation/ld-pruning-wf.md#inputs

juklucas commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,440 @@
+
+		f = open("subset_gds.config", "a")
+		f.write("gds_file " + gds + "\n")
+		f.write("variant_include_file " + variant_include_file + "\n")
```
The CWL has this as optional
```
        	if (inputs.variant_include_file)
        	{
        	    config += 'variant_include_file "' + inputs.variant_include_file.path + "\"\n"
        	}
``` 
We need it to be optional as well.

aofarrel replied:
It's a hard requirement in the CWL, I'm not sure why Stephanie wrote it and the GDS as an if. https://github.com/UW-GAC/analysis_pipeline_cwl/blob/c68ac3f3c8b07512d0fcaffd03ed2d168e294993/relatedness/tools/subset_gds.cwl#L72

juklucas reviewed (changes requested):
Stopped review because I think subset gds needs to be reworked to allow optional inputs for variant_include_file.

aofarrel commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,440 @@
+		import os
+		f = open("ld_pruning.config", "a")
+		f.write('gds_file "~{gds}"\n')
+		if "~{exclude_pca_corr}" != "true":
```
We discussed this on Zoom but it's such a good point I'm going to essentially xpost that discussion on Github and link this to #14.

The CWL's logic seems to the reason behind its apparent bug. Most of the time `if(inputs.parameter)` is essentially the same as checking if something is defined. But because this is a boolean, I think the value of that boolean is influencing it.

In the CWL...
* If the user does not set `exclude_pca_corr` then the expression evaluates to `false` and the input not written to the config.
* If the user sets `exclude_pca_corr` to false then the expression evaluates to `false` and the input is not written to the config. This would result in the R script treating `exclude_pca_corr` as true even though it has been set as false, hence why I'm thinking this is a bug.
* The user is unlikely to set `exclude_pca_corr` to true as that is considered the default value.

aofarrel commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,440 @@
+		File gds
+		File? sample_include_file
+		File? variant_include_file
+		String genome_build = "hg38"  # can also be hg18 or hg19
```
WDL does not have enumerated types, which is why I have to use if/else checks later on down the WDL. I'm down to include parameter_meta{} section but my understanding is that section has no influence on runtime behavior and is just for documentation. I'm also not aware of anything in that section showing up in Terra.

aofarrel replied:
On second thoughts... that section is not used in Terra, and all the information I would include is in the documentation already. It also implies things that aren't necessarily true, ie, if I used the suggestions version for the genome_build parameter then a user may incorrectly assume that variable works like an enumeration rather than a string that is checked on the task level (which would have cost implications).

aofarrel replied:
https://github.com/DataBiosphere/analysis_pipeline_WDL/pull/15#issuecomment-840912964

aofarrel commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,440 @@
+		Float ld_r_threshold = 0.32
+		Float ld_win_size = 10  # yes, a float, not an int
+		Float maf_threshold = 0.01
+		Float missing_threshold = 0.01
```
That would break the python section, as inline python in a WDL cannot detect if something is defined. It'd end up just being a lot of checks of empty strings. I think this method is more explicit and clear for users.

aofarrel replied:
Also, if I'm not mistaken, the CWL has sbdefaultvalue which matches these values.

aofarrel commented:
Mentioned on Zoom:

* Getting rid of overall workflow
* Tidy up structure of the workflow

aofarrel commented:
Wanted to clarify why I'm now against doing parameter_meta in this. Y'see, if we do it here, then that implies we should do it in all upcoming workflows, and uh...

https://github.com/DataBiosphere/analysis_pipeline_WDL/blob/implement-null-model/documentation/null-model-wf.md#additional-null-model-settings-all-optional

juklucas commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,441 @@
+				f.write(chr)
+				f.write(" ")
+			f.write("'")
+			f.write("\ngds_file ")
+			f.write("'")
+			f.write(precisely_one_gds_split[0])  # first part
+			f.write(precisely_one_gds_split[2])  # string "chr"
+			f.write(" ")  # space where R script inserts chr number
+			if(unicode(str(precisely_one_gds_split[1][1])).isnumeric()):
+				# two digit number
+				f.write(precisely_one_gds_split[1][2:])
+			else:
+				# one digit number or Y/X/M
+				f.write(precisely_one_gds_split[1][1:])
+			f.write("'")
```
This should be written on one line using the fact that you already know how to pull the chr. (So you don't need to rewrite that code)

juklucas commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,441 @@
+
+# [4] check_merged_gds -- check a merged GDS file against its inputs
+#
+# This code is nearly identical to the implementation of check_gds in vcf-to-gds-wf.wdl
+# because, like check_gds, you have to remove the chr number of your input file and
+# instead pass it on the command line. It begs the question of "why not just pass in
+# the full name of the file since it's only looking for one file?" but alas it must
+# be passed in this way due to how the R script works.
```
Please remove this comment

juklucas commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,441 @@
+task check_merged_gds {
+	input {
+		File gds
+		File merged
```
The CWL has this as merged_gds_file, we should match that.

juklucas commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,441 @@
+#
+task check_merged_gds {
+	input {
+		File gds
```
The CWL has this as gds_file, we should match that

juklucas commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,441 @@
+
+}
+
+workflow b_ldpruning {
```
Can we just call this ldpruning?

avani-k commented on documentation/checker.md:
```diff
@@ -1,6 +1,8 @@
-This script runs vcf-to-gds-wf.wdl on a known set of inputs, and validates them against expected outputs with a quick md5sum check. On Terra, both inputs and outputs are pulled from the UCSC-owned topmed_workflow_testing bucket.
+This script runs pipelines on a known set of inputs, and validates them against expected outputs with a quick md5sum check. On Terra, both inputs and outputs are pulled from the UCSC-owned topmed_workflow_testing bucket.
 
 This workflow is not simply a copy of the check_gds step. Check_gds would not determine if the unique_variant_ids step ran successfully, instead, it a seperate validation check. Speaking of which, because the inputs are known to not take very long in check_gds, it is enabled *by default.* If you are replacing these files with your own sets of inputs and outputs, and that set is non-downsampled, m ...
```
Check_gds would not determine if the unique_variant_ids step ran successfully, instead, it **<missing verb??>** a **seperate** validation check.

1. Possibly a missing verb for the second half of this sentence?
2. Spelling error for "separate" highlighted in bold.

aofarrel replied:
Thanks for spotting this!

aofarrel commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,441 @@
+				chrom_num = chrom_num.split("chr")
+				chrom_num.append("chr")
+			else:
+				return "error-invalid-inputs"
```
That is the intended behavior.

find_chromosome() is only called once, and it will correctly recognize if that passed in string is valid or not. This same logic is used in vcf-to-gds and I think we've discussed it before. Since it seems to have come up twice, maybe it would be clearer to have it exit within split_on_chromosome() instead?

```
	this_chr = find_chromosome(gds_file)
	if this_chr == "error-invalid-inputs":
		print("Unable to determine chromosome number from inputs.")
		print("Please ensure your files contain ''chr'' followed by")
		print("the number of letter of the chromosome (chr1, chr2, etc)")
		exit(1)
	else:
		chr_array.append(this_chr)
```

juklucas replied:
I don't love this check here, and I didn't like it in the vcf-to-gds wdl either, as you know. By adding it, we are adding a check that the CWL does not have, and that increases the code base to maintain. (It's not a bad check to add, if we are adding stuff, though.)

As for how the split_on_chromosome() should behave, it strikes me as odd throw an error by passing a string to a second function (find_chromosom()) that then ignores it through odd logic, then the original string which is acting like a pseudo-error is passed to main() which THEN recognizes it as the pseudo-error that it is. It would be simpler to just raise an exception in split_on_chromosome(), or you could just exit() in that function.

aofarrel commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,441 @@
+		import os
+
+		def find_chromosome(file):
+			chr_array = []
```
I'm open to that, but the CWL uses arrays.

<img width="1065" alt="Screenshot 2021-05-14 at 1 40 30 PM" src="https://user-images.githubusercontent.com/27784612/118326586-03775100-b4ba-11eb-937c-e78ec79b0003.png">

aofarrel replied:
I changed the logic a bit to make it even closer, let me know if that's what you're looking for.

aofarrel commented:
Up until now, the CWL has been grabbing chromosome numbers by starting from a string that has been split on "chr" and then checking if they are one or two digit by checking if the 1st (0 index) character is numeric.

However, in check_merged_gds the CWL changes tactics. It's instead splitting again, this time on "." and grabbing the 0th result of that split. See [line 63](https://github.com/UW-GAC/analysis_pipeline_cwl/blob/c68ac3f3c8b07512d0fcaffd03ed2d168e294993/relatedness/tools/check_merged_gds.cwl#L63):
`  valueFrom: "${\n    return inputs.gds_file.path.split('chr')[1].split('.')[0]\n}"`

This seems to raise a bit of a dilemma, as our goals for this project may conflict. It's not a matter of difficulty - bringing the current code closer in line to the CWL is simply replacing
 ```
    g = open("chr_number", "a")
    # gds_second_part is already split on 'chr'
    if(unicode(str(gds_second_part[1])).isnumeric()):
	    # two digit number
	    g.write(gds_second_part[:2])
	    gds_second_part = gds_second_part[2:]
    else:
	    # one digit number or Y/X/M
	    g.write(gds_second_part[0])
	    gds_second_part = gds_second_part[1:]
    g.close()
```
with something like `with open("chr_number") write gds_second_part[0].split('.')` or what have you. But that then enforces a new restraint on filenames. We are already restraining users to having `chr` in the filename, but this new function requires it to be set up `like.this.where.there.is.a.period.immediately.after.chrX.where.x.is.the.chromosome.number.gds`

For all TOPMed data I have access to, that's fine, because Gen3 indexes chr-level VCFs to always have a period immediately after the chromosome number. (Yes, I checked all 500ish, what else is there to do on Friday night?) But I only have access to 1000 Genomes + Tutorial in AnVIL Gen3, so I can't check if this would break CCDG or other projects. CCDG has specifically been mentioned as something people want to compare data against when using this pipeline. I also don't know if TOPMed data taken in from outside Gen3 still follows those naming conventions. There could also be other datasets that follow the 'chr' naming convention but not the chr-then-immediate-period convention.

Ultimately, the potential issue at play here is that the AnVIL Project, unlike BioData Catalyst, is mostly WDL and Galaxy focused. That means that this WDL pipeline may have a broader scope than the CWL BDC-and-TOPMed focused version that it is mimicking, so having slightly less strict restraints on input data may be appropriate.

For the time being I propose that the chromosome function is left as it is for now, until we can get clarification from stakeholders as to what are and are not acceptable limitations to place. We can always make the change later.

It could be that chrX + period is the standard across basically everything and I'm writing all this text for nothing, of course, but I don't know that for sure...

aofarrel commented:
Decisions made: 
* Leave the chr number as-is, ie, don't implement the . restriction
* But... don't make this sort of thing a habit

aofarrel commented:
Per [this](https://github.com/DataBiosphere/analysis_pipeline_WDL/pull/15#discussion_r635468084), the 'chr' check is now in the second task. It is not in the first task as the CWL is designed to work with or without chr.

aofarrel commented:
Please look over the changes to the checker workflow (which is now two workflows) and the documentation.

aofarrel commented:
Cost estimates for the workflow + checker will be added once fully non-call cached runs complete on Terra. (Tests already completed successfully on Terra + local, they just had their earlier steps cached.)

juklucas commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,437 @@
+	runtime {
+		cpu: cpu
+		disks: "local-disk " + final_disk_dize + " HDD"
+		docker: "uwgac/topmed-master:2.10.0"
```
At a future date we probably want to set this as a parameter. That way you can give it a default value, but then you can pass different values w/ the checker workflow!

aofarrel replied:
Do you mean setting final_disk_size as a parameter? Right now you can already use the addldisk parameter to throw extra gigabytes into your calculation.

juklucas replied:
No, I meant the Docker container.

juklucas replied:
You are right, putting the container as a parameter is bad practice:
https://docs.dockstore.org/en/develop/advanced-topics/best-practices/best-practices-secure-fair-workflows.html

It looks like the prize for winning the pie eating contest is... more pie. Time to change the Docker container string to include a sha256.

aofarrel replied:
More pie has been added. 🥧

juklucas commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,437 @@
+		File gds_file
+		File merged_gds_file
+
+		# having an input that depends upon a previous task's output reigns in
```
reins in

aofarrel replied:
i really need to rein in my spelling mistakes 😞

juklucas replied:
LOL

juklucas commented on ld-pruning-wf.wdl:
```diff
@@ -0,0 +1,437 @@
+
+		# having an input that depends upon a previous task's output reigns in
+		# cromwell's tendencies to run tasks out of order
+		File? enforce_chronological_order
```
How is this getting used?

aofarrel replied:
In a standard ld-pruning run, not at all. Each task within ld-prune requires an input from the previous task already, so this isn't necessary.

But the checker workflow becomes a chaotic mess without this. Without this enforcement, it may execute in this order:
default case ld prune --> default case subset --> md5 default case subset --> non-default case ld prune --> default case merge -->non-default case subset -->  md5 non-default case subset --> default case check merge --> non-default case merge -->  md5 default case merge --> md5 non-default case merge --> non-default case check merge

Or dozens of other permutations. I want the checker workflow to go in precisely the order I listed the tasks, because if something breaks, it needs to be clear in what step of the original pipeline it would break. If a user has to go through the same "why did the step AFTER the the failure step even execute, much less succeed?" stuff that I did, that's just not a good experience.

juklucas replied:
My question is: you have a parameter in the task, where in *this* task is it getting used?

aofarrel replied:
Nowhere. That's the point. It's a completely bogus file but it manages to trick Cromwell into enforcing an order, because it *thinks* it needs it.

aofarrel replied:
We ultimately decided to remove this from the main workflow and only have the bogus input in the checker.

juklucas reviewed (approved):
I didn't check the documentation, but the ld_pruning wdl looks good. Nice job on that.

aofarrel commented on vcf-to-gds-checker.wdl:
```diff
@@ -0,0 +1,105 @@
+	#            Workflow A            #
+	#           vcf-to-gds-wf          #
+	####################################
+	scatter(wfA_test_vcf in wfA_test_vcfs) {
```
Agreed 100%

aofarrel commented:
Renamed checkers and their respective JSONs passed on Terra.

avani-k commented on documentation/checker.md:
```diff
@@ -1,6 +1,22 @@
-This workflow is not simply a copy of the check_gds step. Check_gds would not determine if the unique_variant_ids step ran successfully, instead, it a seperate validation check. Speaking of which, because the inputs are known to not take very long in check_gds, it is enabled *by default.* If you are replacing these files with your own sets of inputs and outputs, and that set is non-downsampled, m ...
+These workflows are not simply a copy of the check_gds/check_merged_gds steps found in vcf-to-gds and ld-pruning respectively. Instead, they perform a separate validation check using md5 sums.
+
+Due to Cromwell limitations, workflows do not always execute in the order that tasks are written in a WDL file. Efforts were made to reduce in ld-pruning-checker. Similarly, scattered tasks tend to execute their components in a somewhat random order; something like shard-6 --> shard-10 --> shard-20 --> shard-1 is possible even if you have your job concurrency set to 1. For this reason, your pipel ...
```
**Efforts were made to reduce in ld-pruning-checker.**

Can we rephrase this? Efforts were made to reduce what?

aofarrel replied:
It was meant to explain that the checker has some magic that tries to enforce an order, but tbh looking back it's not only unclear but unnecessary information.

avani-k commented on documentation/checker.md:
```diff
@@ -1,6 +1,22 @@
+
+If relying on the truth files located in topmed_workflow_testing, make sure to run this on the full array of input VCF (vcf-to-gds-checker). If you do not run this on the full array of provided files, the resulting outputs will have unexpected variant IDs (vcf-to-gds), resulting in an md5sum mismatch. That does not mean the files are invalid, it just means their IDs are not exactly the same as th ...
+
+ChrX is not in the pipeline's JSON to prevent people who intend on piping the output of this pipeline directly into ld-pruning-checker. However, an input file and truth file are found in the same buckets as the other input and truth buckets for chrX, so you may run this checker on chrX too if you'd like.
```
**ChrX is not in the pipeline's JSON to prevent people who intend on piping the output of this pipeline directly into ld-pruning-checker.**

Can we make this sound a little formal, perhaps in passive tense?

avani-k commented on documentation/cwl-vs-wdl.md:
```diff
@@ -1,24 +1,38 @@
-* The CWL generates config files using an InlineJavascriptRequirement, which is run before the CWL equivivalent of a task's command section begins. The WDL generates them using an inline Python script during the beginning of a task's command section.  
+# ld-pruning-wf.wdl
+## Strictly Necessary Changes
+* Similiar to Workflow A's file relocalization trick in unique_variant_ids, this workflow's merge_gds task has to use the same workaround.
```
I may be confused, but did we decide to get rid of the Workflow A terminology?

aofarrel replied:
Good spot, I missed that one.

avani-k reviewed (approved):
Looks great Ash! Good job! I've added my suggestions.
